File: src/Words.Core/StrTrie.cs
// <copyright file="StrTrie.cs" company="Brian Rogers">
// Copyright (c) Brian Rogers. All rights reserved.
// </copyright>

namespace Words
{
    using System;
    using System.Collections.Generic;
    using System.IO;

    public sealed class StrTrie
    {
        private readonly Dictionary<Str, bool> nodes;

        public StrTrie()
        {
            this.nodes = new Dictionary<Str, bool>();
        }

        public enum NodeKind : byte
        {
            None = 0,
            Prefix = 1,
            Terminal = 2,
        }

        public int Count { get; private set; }

        public static StrTrie Load(Stream stream)
        {
            using (stream)
            {
                StrTrie trie = new StrTrie();
                byte[] buffer = new byte[1024];
                Str value = default(Str);
                bool skip = false;
                int length;
                do
                {
                    length = stream.Read(buffer, 0, buffer.Length);
                    for (int i = 0; i < length; ++i)
                    {
                        char c = (char)buffer[i];
                        switch (c)
                        {
                            case '\r':
                            case '\n':
                                if (!skip && (value.Length > 2))
                                {
                                    trie.Add(value);
                                }

                                value = default(Str);
                                skip = false;
                                break;
                            default:
                                if (!skip)
                                {
                                    if (value.Length == 12)
                                    {
                                        skip = true;
                                    }
                                    else
                                    {
                                        value = value.Append((Ch)(c - 'A' + 1));
                                    }
                                }

                                break;
                        }
                    }
                }
                while (length > 0);

                return trie;
            }
        }

        public void Add(Str value)
        {
            if (value.Length == 0)
            {
                return;
            }

            if (this.nodes.TryGetValue(value, out bool isTerminal) && isTerminal)
            {
                return;
            }

            ++this.Count;
            this.nodes[value] = true;
            while (value.Length > 1)
            {
                value = value.Chop();
                if (this.nodes.ContainsKey(value))
                {
                    return;
                }

                this.nodes.Add(value, false);
            }
        }

        public NodeKind Find(Str value)
        {
            bool isTerminal;
            if (!this.nodes.TryGetValue(value, out isTerminal))
            {
                return NodeKind.None;
            }

            return isTerminal ? NodeKind.Terminal : NodeKind.Prefix;
        }
    }
}

File: src/Words.Native.Core/StrTrie.cpp
#include "StrTrie.h"

using namespace std;
using namespace Words;

StrTrie::StrTrie()
    : nodes_(0.5f), size_(0)
{
}

StrTrie::StrTrie(istream& stream)
    : StrTrie()
{
    char buffer[1024];
    Str value;
    bool skip = false;
    size_t length;
    do
    {
        stream.read(buffer, sizeof(buffer));
        length = stream.gcount();
        for (size_t i = 0; i < length; ++i)
        {
            char c = buffer[i];
            switch (c)
            {
            case '\r':
            case '\n':
                if (!skip && (value.length() > 2))
                {
                    insert(value);
                }

                value = Str();
                skip = false;
                break;
            default:
                if (!skip)
                {
                    if (value.length() == 12)
                    {
                        skip = true;
                    }
                    else
                    {
                        value = value + operator "" _c(c);
                    }
                }

                break;
            }
        }
    } while (length > 0);
}

size_t StrTrie::size() const
{
    return size_;
}

void StrTrie::insert(const Str& value)
{
    if (value.length() == 0)
    {
        return;
    }

    bool b;
    if (!nodes_.insert(value, true, &b))
    {
        if (!b)
        {
            ++size_;
        }

        return;
    }

    ++size_;
    Str v = value;
    while (v.length() > 1)
    {
        v = v.chop();
        bool prev;
        if (!nodes_.insert(v, false, &prev))
        {
            if (prev)
            {
                nodes_.insert(v, true);
            }

            return;
        }
    }
}

StrTrie::NodeKind StrTrie::find(const Str& value) const
{
    bool b;
    if (!nodes_.get(value, b))
    {
        return None;
    }

    return b ? Terminal : Prefix;
}